File: Server/src/main/java/com/github/dr/rwserver/util/LocaleUtil.java
package com.github.dr.rwserver.util;

import com.github.dr.rwserver.dependent.UTF8Control;
import com.github.dr.rwserver.util.log.Log;

import java.text.MessageFormat;
import java.util.Locale;
import java.util.MissingResourceException;
import java.util.ResourceBundle;

/**
 * @author Dr
 * @Date ?
 */
public class LocaleUtil {

    private String[] lg = null;

    public LocaleUtil(String lg) {
        this.lg = lg.split("_");
    }

    private String language(String o,String t,String input,Object[] params) {
        Locale locale = new Locale(o,t);
        ResourceBundle bundle = ResourceBundle.getBundle("bundles/GA", locale, new UTF8Control());
        try {
            if (input !=null){
                if (params == null){
                    return bundle.getString(input);
                }else{
                    return new MessageFormat(bundle.getString(input),locale).format(params);
                }
            }
            //防止使游戏崩溃 CALL..
        } catch (MissingResourceException e) {
            Log.error(e);
        }
        return input+" : Key is invalid.";

    }

    /**
     * 传多参
     * @param      input   语言目标
     * @param      params  替换参
     * @return     文本
     */
    public String getinput(String input,Object... params) {
        if (params == null) {
            return core(input,null);
        }
        Object[] ps = new Object[params.length];
        System.arraycopy(params, 0, ps, 0, params.length);
        return core(input,ps);
    }

    /**
     * 传一数组
     * @param      input  语言目标
     * @param      ps     Object替换组
     * @return     文本
     */
    public String getinputt(String input,Object[] ps) {
        return core(input,ps);
    }

    private String core(String input,Object[] params) {
        String[] lang = lg;
        /*
        if(isBlank(lg)) {
            lang = Config.SERVER_LANGUAGE.split("_");
        }
        */
        if (params == null) {
            return language(lang[0],lang[1],input,null);
        }
        return language(lang[0],lang[1],input,params);
    }
}

File: Server/src/main/java/com/github/dr/rwserver/util/LocaleUtil.kt
/*
 * Copyright 2020-2022 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package com.github.dr.rwserver.util

import com.github.dr.rwserver.struct.OrderedMap
import com.github.dr.rwserver.util.file.FileUtil
import com.github.dr.rwserver.util.io.IoReadConversion
import com.github.dr.rwserver.util.log.Log
import org.jetbrains.annotations.Nls
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.text.MessageFormat
import java.util.*

/**
 * @author Dr
 * @Date ?
 */
class LocaleUtil {
    private val languageData = OrderedMap<String, String>()

    constructor(fileUtil: FileUtil) {
        try {
            load(fileUtil.readInputsStream())
        } catch (e: Exception) {
            Log.fatal("[Load Language Error]", e)
        }
    }

    constructor(inputStream: InputStream) {
        try {
            load(IoReadConversion.fileToReadStream(inputStream))
        } catch (e: Exception) {
            Log.fatal("[Load Language Error]", e)
        }
    }

    @Throws(IOException::class)
    fun addLanguageData(fileUtil: FileUtil) {
        load(fileUtil.readInputsStream())
    }

    private fun load(inputStreamReader: InputStreamReader) {
        try {
            val p = Properties()
            p.load(inputStreamReader)
            p.forEach { key: Any, value: Any -> languageData.put(key as String, value as String) }
        } catch (e: Exception) {
            Log.error("[Load Language File] Error",e)
        }
    }

    private fun language(input: String, params: Array<Any?>?): String {
        val text = languageData[input]
        return if (IsUtil.notIsBlank(text)) {
            if (IsUtil.isBlank(params)) text else MessageFormat(text).format(params)
        } else {
            Log.warn("Translation missing, please check", input)
            "$input : Key is invalid."
        }
    }

    private fun core(input: String, params: Array<Any?>?): String {
        return params?.let { language(input, it) } ?: language(input, null)
    }

    /**
     * 向本地语言中加入自定义
     * @param k String
     * @param v String
     * @return 是否成功
     */
    internal fun addLang(k: String, v: String): Boolean {
        return if (languageData.containsKey(k)) {
            false
        } else {
            languageData.put(k, v)
            true
        }
    }

    /**
     * 传多参
     * @param      input   语言目标
     * @param      params  替换参
     * @return     文本
     */
    @Nls
    fun getinput(input: String, vararg params: Any?): String {
        val ps = arrayOfNulls<Any>(params.size)
        System.arraycopy(params, 0, ps, 0, params.size)
        return core(input, ps)
    }

    /**
     * 传一数组
     * @param      input  语言目标
     * @param      ps     Object替换组
     * @return     文本
     */
    @Nls
    fun getinputt(input: String, ps: Array<Any?>?): String {
        return core(input, ps)
    }


}